include/linux/kfence.h

```diff
@@ -23,31 +22,118 @@ extern char __kfence_pool[KFENCE_POOL_SIZE];
+ * @s: cache being shut down.
+ *
+ * Return: true on success, false if any leftover objects persist.
+ *
+ * Before shutting down a cache, one must ensure there are no remaining objects allocated from it.
+ * KFENCE objects are not referenced from the cache, so kfence_shutdown_cache() takes care of them.
+ */
 bool __must_check kfence_shutdown_cache(struct kmem_cache *s);
 
+
+/**
+ * __kfence_alloc() - allocate a KFENCE object.
+ * @s:     struct kmem_cache with object requirements.
+ * @size:  exact size of the object to allocate (can be less than @s->size e.g. for kmalloc caches).
+ * @flags: GFP flags.
+ *
+ * Return: pointer to a KFENCE object, if available, NULL otherwise.
+ *
+ * __kfence_alloc() performs a deterministic allocation from the KFENCE pool using the requested
+ * parameters.
+ */
 void *__kfence_alloc(struct kmem_cache *s, size_t size, gfp_t flags);
 
-// TODO(elver): Add API doc.
+/**
+ * kfence_alloc() - allocate a KFENCE object with a low probability.
+ * @s:     struct kmem_cache with object requirements.
+ * @size:  exact size of the object to allocate (can be less than @s->size e.g. for kmalloc caches).
+ * @flags: GFP flags.
+ *
+ * Return:
+ * * NULL     - must proceed with allocating from SL[AU]B as usual.
+ * * non-NULL - pointer to a KFENCE object.
+ *
+ * kfence_alloc() should be inserted into the SL[AU]B allocation fast path, allowing it to
+ * transparently return KFENCE-allocated objects with a low probability using a static branch
+ * (the probability is controlled by the kfence.sample_interval boot parameter).
+ */
 static __always_inline void *kfence_alloc(struct kmem_cache *s, size_t size, gfp_t flags)
 {
 	return static_branch_unlikely(&kfence_allocation_key) ? __kfence_alloc(s, size, flags) :
 								      NULL;
 }
 
+/**
+ * kfence_ksize() - get the actual amount of memory allocated for a KFENCE object.
+ * @addr: pointer to a heap object.
+ *
+ * Return:
+ * * 0     - not a KFENCE object, must call __ksize() instead.
+ * * non-0 - this many bytes can be accessed without causing a memory error.
+ *
+ * ksize() returns the number of bytes past the end of a SL[AU]B object that can be safely used
+ * (e.g. because of a generous alignment). kfence_ksize() mimics this behavior for KFENCE objects.
+ */
 size_t kfence_ksize(const void *addr);
 
+/**
+ * kfence_object_start() - find the beginning of a KFENCE object.
+ * @addr - address within a KFENCE-allocated object.
+ *
+ * Return: address of the beginning of the object.
+ *
+ * SL[AU]B-allocated objects are laid out within a page one by one, so it is easy to calculate the
+ * beginning of an object given a pointer inside it and the object size. The same is not true for
+ * KFENCE, which places a single object at either end of the page. This helper function is used to
+ * find the beginning of a KFENCE-allocated object.
+ */
 void *kfence_object_start(const void *addr);
 
+/**
+ * __kfence_free() - release a KFENCE-allocated object to KFENCE pool.
```

Comment:
Like above, I'm not sure we should add official documentation for `__kfence_free()`, because there is no stub. I would still add a non-doc comment, saying what it does but then also saying it should *not* be used and allocators should use kfence_free().

Reply:
Done.